src/platform/datapath_raw_socket.c

```diff
@@ -849,21 +1127,188 @@ CxPlatFramingUdpChecksum(
+        if (SendData == NULL) {
+            return;
+        }
+
+        QuicTraceEvent(
+            DatapathSend,
+            "[data][%p] Send %u bytes in %hhu buffers (segment=%hu) Dst=%!ADDR!, Src=%!ADDR!",
+            Socket,
+            SendData->Buffer.Length,
+            1,
+            (uint16_t)SendData->Buffer.Length,
+            CASTED_CLOG_BYTEARRAY(sizeof(Route->RemoteAddress), &Route->RemoteAddress),
+            CASTED_CLOG_BYTEARRAY(sizeof(Route->LocalAddress), &Route->LocalAddress));
+        CxPlatFramingWriteHeaders(
+            Socket, Route, &SendData->Buffer, SendData->ECN,
+            Interface->OffloadStatus.Transmit.NetworkLayerXsum,
+            Interface->OffloadStatus.Transmit.TransportLayerXsum, 
+            ReceivedTcpHeader->AckNumber,
+            CxPlatByteSwapUint32(CxPlatByteSwapUint32(ReceivedTcpHeader->SequenceNumber) + 1),
+            TH_RST | TH_ACK);
+        Socket->CachedRstSend = SendData;
+    }
+}
+
+_IRQL_requires_max_(DISPATCH_LEVEL)
+void
+CxPlatDpRawSocketSyn(
+    _In_ CXPLAT_SOCKET* Socket,
+    _In_ const CXPLAT_ROUTE* Route
+    )
+{
+    CXPLAT_DBG_ASSERT(Socket->UseTcp);
+
+    CXPLAT_SEND_CONFIG SendConfig = { (CXPLAT_ROUTE*)Route, 0, CXPLAT_ECN_NON_ECT, 0 };
+    CXPLAT_SEND_DATA *SendData = CxPlatSendDataAlloc(Socket, &SendConfig);
+    if (SendData == NULL) {
+        return;
+    }
+
+    QuicTraceEvent(
+        DatapathSend,
+        "[data][%p] Send %u bytes in %hhu buffers (segment=%hu) Dst=%!ADDR!, Src=%!ADDR!",
+        Socket,
+        SendData->Buffer.Length,
+        1,
+        (uint16_t)SendData->Buffer.Length,
+        CASTED_CLOG_BYTEARRAY(sizeof(Route->RemoteAddress), &Route->RemoteAddress),
+        CASTED_CLOG_BYTEARRAY(sizeof(Route->LocalAddress), &Route->LocalAddress));
+    CXPLAT_DBG_ASSERT(Route->State == RouteResolved);
+    CXPLAT_DBG_ASSERT(Route->Queue != NULL);
+    const CXPLAT_INTERFACE* Interface = CxPlatDpRawGetInterfaceFromQueue(Route->Queue);
+    CxPlatFramingWriteHeaders(
+        Socket, Route, &SendData->Buffer, SendData->ECN,
+        Interface->OffloadStatus.Transmit.NetworkLayerXsum,
+        Interface->OffloadStatus.Transmit.TransportLayerXsum, 
+        Route->TcpState.SequenceNumber, 0, TH_SYN);
+    CxPlatDpRawTxEnqueue(SendData);
 }
 
 _IRQL_requires_max_(DISPATCH_LEVEL)
 void
 CxPlatFramingWriteHeaders(
-    _In_ const CXPLAT_SOCKET* Socket,
+    _In_ CXPLAT_SOCKET* Socket,
     _In_ const CXPLAT_ROUTE* Route,
     _Inout_ QUIC_BUFFER* Buffer,
     _In_ CXPLAT_ECN_TYPE ECN,
     _In_ BOOLEAN SkipNetworkLayerXsum,
-    _In_ BOOLEAN SkipTransportLayerXsum
+    _In_ BOOLEAN SkipTransportLayerXsum,
+    _In_ uint32_t TcpSeqNum,
+    _In_ uint32_t TcpAckNum,
+    _In_ uint8_t TcpFlags
     )
 {
-    UDP_HEADER* UDP = (UDP_HEADER*)(Buffer->Buffer - sizeof(UDP_HEADER));
+    uint8_t *Transport;
+    uint16_t TransportLength;
+    uint8_t TransportProtocol = Socket->UseTcp ? IPPROTO_TCP : IPPROTO_UDP;
```

Comment:
Move the set of this down into the if/else below so we don't have to do the branch twice.